fuzz: setup memcs fuzzing

drewdzzz opened the pull request:
The patchset adapts engine fuzzing to `memcs` and populates it with memcs-specific operations. Along the way, the fuzzing is populated with fixed-sized types and nullable fields that are commonly used by `memcs`.

[**EE part**](https://github.com/tarantool/tarantool-ee/pull/981) - only adapts `arrow_c_api_wrapper` to the fuzzing test.

Part of tarantool/tarantool-ee#834

ligurio commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -113,9 +132,17 @@ local err_pat_whitelist = {
+    "Index '[%w_]+' %(TREE%) of space '[%w_]+' %(memcs%) does not support pagination",
+    "Arrow stream does not support field type '[%w_]+'",
+    "box_insert_arrow: field [%d]+ has unsupported type",
+    "Engine 'memcs' does not support upsert%(%)",
+    "Engine 'memcs' does not support index build",
+    "Engine 'memcs' does not support secondary indexes",
+    "Engine 'memcs' does not support variable field count",
```
Isn't better to avoid using unsupported operations?

drewdzzz replied:
Makes sense. I will add a new `engines` attributes to filter the operations.

ligurio replied:
Now engine is checked in `insert_arrow_op`, `scan_column_op`, `arrow_stream_op`, but memcs-specific errors are still in `err_pat_whitelist`.

drewdzzz replied:
Added the new attribute `engines` to the list of the ops.

ligurio reviewed (commented):
Andrey, thanks for your patches! Please see my comments.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -485,10 +582,87 @@ local function format_op(space, space_format)
+            log.info('SCAN_COLUMN: finish early')
+            break
+        end
+        local max_batch_size = oneof({10, 10^6})
```
I choose here between a small batch (up to 10 elements) and a large batch (up to 10^6 elements) so that small batches appear much more frequently. If I used `math.random(10^6)`, there would be only `1/10^5` probability of a small batch.

ligurio replied:
please add a comment

drewdzzz replied:
Done.

ligurio replied:
Thanks!

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -485,10 +582,87 @@ local function format_op(space, space_format)
+        end
+        local max_batch_size = oneof({10, 10^6})
+        local batch_size = math.random(1, max_batch_size)
+        log.info('SCAN_COLUMN: going to read batch of size ' .. batch_size)
```
I consider each advance of scanner column as a separate operation so I've deliberately chosen `info` level. Do you think it's better to consider it as a one large operation and do not log each advancement manually?

By the way, debug log info seems to be useless - the fuzzing is not actually stable in the sense of reproducing crashes having a seed (at least from my experience).

ligurio replied:
> the fuzzing is not actually stable in the sense of reproducing crashes having a seed

The test is absolutely deterministic and everything used in test depends on a seed, the problem is in non-deterministic Tarantool.

> By the way, debug log info seems to be useless

Ok, let's leave with `info` verbosity if it is more convenient for you.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -788,6 +991,46 @@ local function random_key(space, idx)
+    local format_size = #space:format()
+    local batch = {}
+    -- Randomly choose between small and big batch
+    local max_batch_size = oneof({10, 100})
```
> Is it intentional?

Yep, here is the reason. We pass the generated batch to the INSERT_ARROW_OP and it is logged unconditionally from its size. Hence, too large batches will clutter up the logs. In `arrow_stream_op` function I log the batch manually so I can log only small batches.

Probably, we should automatically limit the length of log message so that I could generate huge batches here as well without cluttering logs.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -279,6 +280,70 @@ local tarantool_type = {
         generator = uuid.new,
         operations = {'=', '!'},
     },
```
Oh, thanks! It seems we should introduce a new attribute like `since` for the types to specify Tarantool version.

ligurio replied:
What are the changes?

drewdzzz replied:
What's the policy of backward compatibility of the fuzzer?
I tried to run actual version of fuzzer (took it from master) for Tarantool 2.11.5 and it just failed to start:
```
ER_CFG: Incorrect value for option 'memtx_sort_threads': unexpected option
```

ligurio replied:
The test is not executed automatically in CI, so it seems it was broken for 2.11. According to this and because 2.11 is not in active development, I propose to omit support of 2.11 by the test and focus only on 3.x.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -112,9 +134,20 @@ local err_pat_whitelist = {
+    "Can't modify space '[%d]+': the space was concurrently modified",
+    "Failed to write to disk",
+    "WAL has a rollback in progress",
+    -- MEMCS
```
Fixed.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -985,6 +1228,26 @@ local ops = {
         args = function(_) return end,
     },
+
+    -- MEMCS
```
Fixed.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -15,14 +15,18 @@ Usage: tarantool test_engine.lua
 local log = require('log')
 local math = require('math')
+local column_scanner = nil
+local arrow = nil
```
Fixed.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -119,6 +119,8 @@ local err_pat_whitelist = {
     -- DDL on a space is locked until the end of the current DDL
```
Rewrote the commit message - PTAL =)

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -99,6 +119,8 @@ local err_pat_whitelist = {
     -- DDL on a space is locked until the end of the current DDL
     -- operation.
     "the space is already being modified",
+    -- The test actively uses error injections that can cause such errors
```
Fixed all the places with a missing dot.

drewdzzz commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -85,6 +89,22 @@ local seed = params.seed or os.time()
+    if tarantool.package ~= 'Tarantool Enterprise' then
+        error('Engine ' .. arg_engine .. ' requires Tarantool Enterprise')
+    end
+    local mod_pattern = '?.' .. tarantool.build.mod_format
+    local test_dir = fio.abspath(
+        fio.pathjoin(os.getenv('BUILDDIR') or '.',
+        'test', 'enterprise-luatest'))
+    local test_mod_cpath = fio.pathjoin(test_dir, mod_pattern)
+    package.cpath = test_mod_cpath .. ';' .. package.cpath
```
Do you mean introducing a new option to the script? (Like `tarantool test_engine.lua --lua_cpath <path>`)

ligurio replied:
> Do you mean introducing a new option to the script? (Like tarantool test_engine.lua --lua_cpath <path>)

No.

Imagine `LUA_PATH` and `LUA_CPATH` are passed outside and everything you need is require modules.

```diff
@@ -89,18 +89,13 @@ local seed = params.seed or os.time()
 math.randomseed(seed)
 log.info('Random seed: %d', seed)
 
--- Set package.cpath to allow loading C modules from the test directory.
+-- MEMCS engine requires two Lua C modules:
+-- `arrow_c_api_wrapper` and `scanner_c_api_wrapper`.
 if arg_engine == 'memcs' then
     local tarantool = require('tarantool')
     if tarantool.package ~= 'Tarantool Enterprise' then
         error('Engine ' .. arg_engine .. ' requires Tarantool Enterprise')
     end
-    local mod_pattern = '?.' .. tarantool.build.mod_format
-    local test_dir = fio.abspath(
-        fio.pathjoin(os.getenv('BUILDDIR') or '.',
-        'test', 'enterprise-luatest'))
-    local test_mod_cpath = fio.pathjoin(test_dir, mod_pattern)
-    package.cpath = test_mod_cpath .. ';' .. package.cpath
     column_scanner = require('scanner_c_api_wrapper')
     arrow = require('arrow_c_api_wrapper')
 end
```

and then: ` LUA_CPATH='build/test/enterprise-luatest/%.so;;' tarantool test/fuzz/lua/test_engine.lua`.
It's easier to set such env variables in CMake then construct in test itself.

drewdzzz replied:
Done.

ligurio commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -485,10 +582,87 @@ local function format_op(space, space_format)
+        return
+    end
+    local mpkey = msgpack.encode(key)
+    local batch_size = nil
```
It is a `nil` by default.

drewdzzz replied:
Thanks, fixed.

ligurio commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -485,10 +582,87 @@ local function format_op(space, space_format)
+    local mpkey = msgpack.encode(key)
+    local scanner = column_scanner.box_index_scanner(space.id, 0, fields, mpkey)
+    while true do
+        if math.random(1, 20) == 1 then
```
please add a comment

drewdzzz replied:
Done.

ligurio replied:
Thanks!

drewdzzz commented:
The PR was massively reworked:
1. There are several new commits fixing some minor problems of the fuzzer.
2. Added support of secondary memcs indexes.
3. Whitelist of all errors is updated accordingly to new behavior of cross-engine transactions.

ligurio commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -500,12 +621,23 @@ local function format_op(space, space_format)
     -- Configuration reference (box.cfg),
     -- https://www.tarantool.io/en/doc/latest/reference/configuration/
+
+    -- vinyl_page_size must be in range (0, vinyl_range_size].
```
please move comment under the line `local vinyl_page_size = ...`

drewdzzz replied:
Done.

ligurio commented on test/fuzz/lua/test_engine.lua:
```diff
@@ -848,6 +1157,7 @@ local ops = {
         args = function(space)
             return random_tuple(space:format()), random_tuple_operations(space)
         end,
+        engines = {'memtx', 'vinyl'},
```
Do we really need additional field if it is required in a single operation?
I propose to infer engine name using index and return from an operation function if engine is not supported:

```lua

tarantool> box.space.aa.index['pk'].space_id
---
- 512
...

tarantool> box.space[512].engine
---
- memtx
...

tarantool> 
```

What do you think?

ligurio replied:
Okay, thanks! (Actually, I've added a comment on review an intermediate commit when didn't see the latest commit where more engines limitation were added.)

ligurio commented:
Andrey, thanks for your patches! I left a number of comments, but most of them are minors. Let's converge with review and merge all these changes to the master.

**fuzz: respect constraints of vinyl_page_size configuration parameter**

Minor: Short message is longer 50 symbols, long message is longer 72 symbols. Refer to Developer guidelines [^1].

**fuzz: disable min and max for rtree indexes**

Minor: Long message is longer 72 symbols. Refer to Developer guidelines [^1].
BTW, it is not described in documentation that `min()` and `max()` are not supported for RTree, neither in [^2] nor in [^3].

Minor: s/rtree/RTREE/ and s/Rtree/RTREE.

**fuzz: enable INDEX_RANDOM_OP for memtx tree**

Minor: s/tree/TREE/.

**fuzz: populate engine fuzzing with nullable fields**

please add a link to doc about nullable fields (https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/create_index/#key-part-is-nullable)

**fuzz: populate engine fuzzing with fixed-size types**

Minor: Short message is longer 50 symbols, long message is longer 72 symbols. Refer to Developer guidelines [^1].

And it is a good idea with supporting other types, thanks!

**fuzz: update suppression list**

Please add `Follows up #1803` to the commit message.

[^1]: https://www.tarantool.io/en/doc/2.11/dev_guide/developer_guidelines/#how-to-write-a-commit-message
[^2]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/max/
[^3]: https://www.tarantool.io/en/doc/latest/platform/ddl_dml/indexes/#indexes-rtree

drewdzzz commented:
Thank you for your review, I fixed all the notions.

You asked me to refer to the documentation of `is_nullable` attribute. I placed the link next to `tarantool_indices` table - there was `TODO: support 'is_nullable'` comment before my patch.

<details>

<summary> diff </summary>

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index 5c14ed3b57..f69437d6cc 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -400,6 +400,8 @@ end
 
 -- Iterator types for indexes.
 -- See https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/pairs/#box-index-iterator-types
+-- For information about nullable fields, see:
+-- https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/create_index/#lua-data.key_part.is_nullable
 -- TODO: support `multikey`.
 -- TODO: support `pagination`.
 local tarantool_indices = {
```
</details>

Actually, nullable fields didn't work because I didn't set this attribute in space format. I fixed the problem, added a new error to whitelist and found out that BITSET and HASH are not nullable as well.
<details>
<summary>diff</summary>

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index 63c3a31a22..21a71d443f 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -118,6 +118,8 @@ local err_pat_whitelist = {
     "Failed to allocate %d+ bytes in [%w_]+ for [%w_]+",
     "Storage engine 'memtx' does not support cross%-engine transactions",
     "Storage engine 'vinyl' does not support cross%-engine transactions",
+    "Tuple field %d+ %([%w_]+%) type does not match one required by " ..
+        "operation: expected %w+, got nil",
 }
 
 local function keys(t)
@@ -382,12 +384,14 @@ local function random_space_format()
         table.insert(space_format, {
             name =('field_%d'):format(i),
             type = datatype,
+            is_nullable = oneof({true, false}),
         })
     end
     for i = min_num_fields - 1, num_fields - min_num_fields - 1 do
         table.insert(space_format, {
             name =('field_%d'):format(i),
             type = oneof(keys(tarantool_type)),
+            is_nullable = oneof({true, false}),
         })
     end
 
@@ -419,7 +423,7 @@ local tarantool_indices = {
         is_multipart = true,
         is_min_support = false,
         is_max_support = false,
-        is_nullable_support = true,
+        is_nullable_support = false,
         is_unique_support = true,
         is_non_unique_support = false,
         is_primary_key_support = true,
@@ -441,7 +445,7 @@ local tarantool_indices = {
         is_multipart = false,
         is_min_support = false,
         is_max_support = false,
-        is_nullable_support = true,
+        is_nullable_support = false,
         is_unique_support = false,
         is_non_unique_support = true,
         is_primary_key_support = false,
```
</details>

Also, memcs doesn't support nullable fields for now. Fixed the problem as well:

<details>

<summary> diff </summary>

```diff
@@@ -720,6 -764,6 +768,7 @@@ local function index_opts(space, is_pri
      local n_parts = math.random(1, table.getn(possible_fields))
      local id = unique_ids(n_parts)
      local is_nullable_support = not is_primary and
+        space.engine ~= 'memcs' and
          tarantool_indices[opts.type].is_nullable_support
      for i = 1, n_parts do
          local field_id = id()
```
</details>

Another problem revealed by fixed nullable fields - we should use index id instead of its key in `space.index` to determine if it is primary because the key can be the index name. Fixed in the commit introducing nullable fields, commit message was updated correspondingly.

<details>

<summary> diff </summary>

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index 25a33872cf..e64f22ee45 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -1182,8 +1182,9 @@ local ops = {
         func = index_alter_op,
         args = function(space)
             local idx_n = oneof(keys(space.index))
-            local is_primary = idx_n == 0
-            return space.index[idx_n], index_opts(space, is_primary)
+            local index = space.index[idx_n]
+            local is_primary = index.id == 0
+            return index, index_opts(space, is_primary)
         end,
     },
     INDEX_COMPACT_OP = {
```
</details>

ligurio commented:
**fuzz: respect constraints of vinyl_page_size configuration parameter**

please add a line `Follows up commit 2fa1ca66eb40 ("test/fuzz: adjust vinyl settings").` to the commit message.

ligurio reviewed (approved):
Andrey, thanks for your patches!
